File: app/src/main/java/org/adaway/model/backup/BackupImporter.java
package org.adaway.model.backup;

import android.content.Context;
import android.net.Uri;
import android.widget.Toast;

import androidx.annotation.UiThread;

import org.adaway.R;
import org.adaway.db.AppDatabase;
import org.adaway.db.dao.HostListItemDao;
import org.adaway.db.dao.HostsSourceDao;
import org.adaway.db.entity.HostListItem;
import org.adaway.db.entity.ListType;
import org.adaway.util.AppExecutors;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Optional;
import java.util.concurrent.Executor;

import static org.adaway.db.entity.ListType.ALLOWED;
import static org.adaway.db.entity.ListType.BLOCKED;
import static org.adaway.db.entity.ListType.REDIRECTED;
import static org.adaway.model.backup.BackupFormat.ALLOWED_KEY;
import static org.adaway.model.backup.BackupFormat.BLOCKED_KEY;
import static org.adaway.model.backup.BackupFormat.REDIRECTED_KEY;
import static org.adaway.model.backup.BackupFormat.SOURCES_KEY;
import static org.adaway.model.backup.BackupFormat.hostFromJson;
import static org.adaway.model.backup.BackupFormat.sourceFromJson;

import timber.log.Timber;

/**
 * This class is a helper class to import user lists and hosts sources to a backup file.<br>
 * Importing a file source will no restore read access from Storage Access Framework.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
public final class BackupImporter {

    private BackupImporter() {

    }

    private static final Executor DISK_IO_EXECUTOR = AppExecutors.getInstance().diskIO();
    private static final Executor MAIN_THREAD_EXECUTOR = AppExecutors.getInstance().mainThread();

    /**
     * Import a backup file.
     *
     * @param context   The application context.
     * @param backupUri The URI of a backup file.
     */
    @UiThread
    public static void importFromBackup(Context context, Uri backupUri) {
        DISK_IO_EXECUTOR.execute(() -> {
            boolean imported = true;
            try {
                importBackup(context, backupUri);
            } catch (IOException e) {
                Timber.e(e, "Failed to import backup.");
                imported = false;
            }
            boolean successful = imported;
            MAIN_THREAD_EXECUTOR.execute(() -> notifyImportEnd(context, successful));
        });
    }

    @UiThread
    private static void notifyImportEnd(Context context, boolean successful) {
        Toast.makeText(
                context,
                context.getString(successful ? R.string.import_success : R.string.import_failed),
                Toast.LENGTH_LONG
        ).show();
    }

    static void importBackup(Context context, Uri backupUri) throws IOException {
        try (InputStream inputStream = context.getContentResolver().openInputStream(backupUri);
             BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
            StringBuilder contentBuilder = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                contentBuilder.append(line);
            }
            JSONObject backupObject = new JSONObject(contentBuilder.toString());
            importBackup(context, backupObject);
        } catch (JSONException exception) {
            throw new IOException("Failed to parse backup file.", exception);
        } catch (FileNotFoundException exception) {
            throw new IOException("Failed to find backup file.", exception);
        } catch (IOException exception) {
            throw new IOException("Failed to read backup file.", exception);
        }
    }

    private static void importBackup(Context context, JSONObject backupObject) throws JSONException {
        AppDatabase database = AppDatabase.getInstance(context);
        HostsSourceDao hostsSourceDao = database.hostsSourceDao();
        HostListItemDao hostListItemDao = database.hostsListItemDao();

        importSourceBackup(hostsSourceDao, backupObject.getJSONArray(SOURCES_KEY));
        importListBackup(hostListItemDao, BLOCKED, backupObject.getJSONArray(BLOCKED_KEY));
        importListBackup(hostListItemDao, ALLOWED, backupObject.getJSONArray(ALLOWED_KEY));
        importListBackup(hostListItemDao, REDIRECTED, backupObject.getJSONArray(REDIRECTED_KEY));
    }

    private static void importSourceBackup(HostsSourceDao hostsSourceDao, JSONArray sources) throws JSONException {
        for (int index = 0; index < sources.length(); index++) {
            JSONObject sourceObject = sources.getJSONObject(index);
            hostsSourceDao.insert(sourceFromJson(sourceObject));
        }
    }

    private static void importListBackup(HostListItemDao hostListItemDao, ListType type, JSONArray hosts) throws JSONException {
        for (int index = 0; index < hosts.length(); index++) {
            JSONObject hostObject = hosts.getJSONObject(index);
            HostListItem host = hostFromJson(hostObject);
            host.setType(type);
            Optional<Integer> id = hostListItemDao.getHostId(host.getHost());
            if (id.isPresent()) {
                host.setId(id.get());
                hostListItemDao.update(host);
            } else {
                hostListItemDao.insert(host);
            }
        }
    }
}

File: app/src/main/java/org/adaway/model/backup/BackupImporter.kt
package org.adaway.model.backup

import android.content.Context
import android.net.Uri
import android.widget.Toast
import androidx.annotation.UiThread
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.adaway.R
import org.adaway.db.AppDatabase
import org.adaway.db.entity.HostListItem
import org.adaway.db.entity.ListType
import org.adaway.db.entity.ListType.ALLOWED
import org.adaway.db.entity.ListType.BLOCKED
import org.adaway.db.entity.ListType.REDIRECTED
import org.adaway.model.backup.BackupFormat.ALLOWED_KEY
import org.adaway.model.backup.BackupFormat.BLOCKED_KEY
import org.adaway.model.backup.BackupFormat.REDIRECTED_KEY
import org.adaway.model.backup.BackupFormat.SOURCES_KEY
import org.adaway.model.backup.BackupFormat.hostFromJson
import org.adaway.model.backup.BackupFormat.sourceFromJson
import org.adaway.util.ExpressiveToast
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStreamReader

object BackupImporter {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    @JvmStatic
    @UiThread
    fun importFromBackup(context: Context, backupUri: Uri) {
        scope.launch {
            val imported = withContext(Dispatchers.IO) {
                try {
                    importBackup(context, backupUri)
                    true
                } catch (exception: IOException) {
                    Timber.e(exception, "Failed to import backup.")
                    false
                }
            }
            notifyImportEnd(context, imported)
        }
    }

    @UiThread
    private fun notifyImportEnd(context: Context, successful: Boolean) {
        ExpressiveToast.makeText(
            context,
            context.getString(if (successful) R.string.import_success else R.string.import_failed),
            Toast.LENGTH_LONG
        ).show()
    }

    @JvmStatic
    @Throws(IOException::class)
    fun importBackup(context: Context, backupUri: Uri) {
        try {
            val inputStream = context.contentResolver.openInputStream(backupUri)
                ?: throw FileNotFoundException("Could not open backup file.")
            val content = inputStream.use { stream ->
                BufferedReader(InputStreamReader(stream)).use { reader ->
                    buildString {
                        var line = reader.readLine()
                        while (line != null) {
                            append(line)
                            line = reader.readLine()
                        }
                    }
                }
            }
            importBackup(context, JSONObject(content))
        } catch (exception: JSONException) {
            throw IOException("Failed to parse backup file.", exception)
        } catch (exception: FileNotFoundException) {
            throw IOException("Failed to find backup file.", exception)
        } catch (exception: IOException) {
            throw IOException("Failed to read backup file.", exception)
        }
    }

    @Throws(JSONException::class)
    private fun importBackup(context: Context, backupObject: JSONObject) {
        val database = AppDatabase.getInstance(context)
        val hostsSourceDao = database.hostsSourceDao()
        val hostListItemDao = database.hostsListItemDao()

        importSourceBackup(hostsSourceDao, backupObject.getJSONArray(SOURCES_KEY))
        importListBackup(hostListItemDao, BLOCKED, backupObject.getJSONArray(BLOCKED_KEY))
        importListBackup(hostListItemDao, ALLOWED, backupObject.getJSONArray(ALLOWED_KEY))
        importListBackup(hostListItemDao, REDIRECTED, backupObject.getJSONArray(REDIRECTED_KEY))
    }

    @Throws(JSONException::class)
    private fun importSourceBackup(
        hostsSourceDao: org.adaway.db.dao.HostsSourceDao,
        sources: JSONArray
    ) {
        for (index in 0 until sources.length()) {
            hostsSourceDao.insert(sourceFromJson(sources.getJSONObject(index)))
        }
    }

    @Throws(JSONException::class)
    private fun importListBackup(
        hostListItemDao: org.adaway.db.dao.HostListItemDao,
        type: ListType,
        hosts: JSONArray
    ) {
        for (index in 0 until hosts.length()) {
            val host: HostListItem = hostFromJson(hosts.getJSONObject(index))
            host.type = type
            val id = hostListItemDao.getHostId(host.host)
            if (id.isPresent) {
                host.id = id.get()
                hostListItemDao.update(host)
            } else {
                hostListItemDao.insert(host)
            }
        }
    }
}
